Change demo/meeting deployment to CDK

schuettc opened the pull request:
**Issue #:** 
Replace SAM deployment of demo/meeting with CDK deployment

**Testing**
1. Have you successfully run `npm run build:release` locally?

Yes

2. How did you test these changes?
Locally

3. If you made changes to the component library, have you provided corresponding documentation changes?
N/A

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
Confirm (Also agree to MIT-0)

devalevenkatesh commented on demo/meeting/back-end-resources/lambda-layer/nodejs/node_modules/utils.js:
```diff
@@ -0,0 +1,82 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0var AWS = require('aws-sdk');
+
+var AWS = require('aws-sdk');
```
Should we have this `util.js` under the `lambda-layer` or `nodejs` than the `node_modules`?

schuettc replied:
That's where it has to go.  The directory structure for layers requires it.

devalevenkatesh reviewed (commented):
Can we keep all the front-end as is and just change the backend deployment, so deploy what webpack generates using CDK and not SAM?

I see we are adding lot of plugins and node environment changes here which looks a lot complex than what currently exists. Adding these plugins will make it hard to maintain right? Also, we do not want to have any dependency on the `react-dev-utils` as well if we can.

michhyun1 commented on demo/meeting/front-end-resources/react-meeting/package.json:
```diff
@@ -0,0 +1,199 @@
+    "@typescript-eslint/eslint-plugin": "^4.5.0",
+    "@typescript-eslint/parser": "^4.5.0",
+    "amazon-chime-sdk-component-library-react": "file:.yalc/amazon-chime-sdk-component-library-react",
+    "amazon-chime-sdk-js": "^2.3.0",
```
We're on 2.7.0 now FYI

michhyun1 commented on demo/meeting/front-end-resources/react-meeting/package.json:
```diff
@@ -0,0 +1,199 @@
+    "@typescript-eslint/parser": "^4.5.0",
+    "amazon-chime-sdk-component-library-react": "file:.yalc/amazon-chime-sdk-component-library-react",
+    "amazon-chime-sdk-js": "^2.3.0",
+    "aws-amplify": "^3.3.13",
```
are we using amplify?

schuettc replied:
Cleaned up package.json

michhyun1 commented on demo/meeting/front-end-resources/react-meeting/package.json:
```diff
@@ -0,0 +1,199 @@
+    "ts-pnp": "1.2.0",
+    "typescript": "^4.1.3",
+    "url-loader": "4.1.1",
+    "web-vitals": "^0.2.4",
```
Do we use this?

michhyun1 commented on demo/meeting/front-end-resources/react-meeting/package.json:
```diff
@@ -0,0 +1,199 @@
+    "eslint-plugin-react": "^7.20.5",
+    "eslint-plugin-react-hooks": "^1.7.0",
+    "html-webpack-inline-source-plugin": "0.0.10",
+    "html-webpack-plugin": "^3.2.0",
```
this is a devDependency and Dependency?

michhyun1 commented on demo/meeting/front-end-resources/react-meeting/package.json:
```diff
@@ -0,0 +1,199 @@
+      "<rootDir>/src/**/*.{spec,test}.{js,jsx,ts,tsx}"
+    ],
+    "testEnvironment": "jsdom",
+    "testRunner": "/Users/courtschuett/Documents/Projects/amazon-chime-sdk-component-library-react-fork/demo/meeting/front-end-resources/react-meeting/node_modules/jest-circus/runner.js",
```
Nit: Need to update

michhyun1 commented on demo/meeting/front-end-resources/react-meeting/src/containers/DevicePermissionPrompt.tsx:
```diff
@@ -0,0 +1,51 @@
+// // Copyright 2020-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// // SPDX-License-Identifier: Apache-2.0
+
+// import React from 'react';
+
+// import {
+//   Modal,
+//   ModalBody,
+//   ModalHeader,
+//   DevicePermissionStatus,
+//   useDevicePermissionStatus
+// } from 'amazon-chime-sdk-component-library-react';
+
+// import Card from '../components/Card';
+
+// // Show permission prompt when the user is granting the browser permissions
+// // Show nothing if permission is already granted or component loads on initial render
```
Why is this commented out?

schuettc replied:
This should be added back in.

michhyun1 commented on demo/meeting/front-end-resources/react-meeting/src/containers/MeetingForm/index.tsx:
```diff
@@ -19,7 +19,7 @@ import { getErrorContext } from '../../providers/ErrorProvider';
 import Card from '../../components/Card';
 import Spinner from '../../components/Spinner';
-import DevicePermissionPrompt from '../DevicePermissionPrompt';
+// import DevicePermissionPrompt from '../DevicePermissionPrompt';
```
Why arent we using this?

schuettc replied:
Added back in

michhyun1 commented on demo/meeting/front-end-resources/react-meeting/src/constants/routes.ts:
```diff
@@ -0,0 +1,25 @@
+  DEVICE: `${rootPath}devices`,
+  MEETING: `${rootPath}meeting`,
+  JOIN: `${rootPath}join`,
+  JOIN_BREAKOUT: `${rootPath}joinbreakout`
```
Where do we use this?

schuettc replied:
Breakout references should be removed.

schuettc commented on demo/meeting/back-end-resources/back-end.ts:
```diff
@@ -0,0 +1,196 @@
+    
+        lambdaChimeRole.addToPolicy(new iam.PolicyStatement({
+          resources: ['*'],
+          actions: ['chime:CreateMeeting',
+                    'chime:CreateMeeting',
+                    'chime:CreateMeeting',
+                    'chime:DeleteMeeting',
+                    'chime:GetMeeting',
+                    'chime:ListMeetings',
+                    'chime:BatchCreateAttendee',
+                    'chime:CreateAttendee',
+                    'chime:DeleteAttendee',
+                    'chime:GetAttendee',
+                    'chime:ListAttendees',
```
This can likely be tuned better.  I just took them from the original: https://github.com/aws/amazon-chime-sdk-component-library-react/blob/091d371167145bd95aaac68848b7abcf9dce3499/demo/meeting/serverless/template.yaml#L33

michhyun1 replied:
Maybe we can just remove the duplicate CreateMeetings then

schuettc replied:
Done.

schuettc commented on demo/meeting/deploy.sh:
```diff
@@ -0,0 +1,48 @@
+    echo ""
+    npm install --legacy-peer-deps
+fi
+if [ ! -d "front-end-resources/react-meeting/build" ]; then
+    echo ""
+    echo "Creating front-end-resources/react-meeting/build directory"
+    echo ""
+    mkdir front-end-resources/react-meeting/build
+fi
```
Possibly, yes.  I can try without this.

schuettc replied:
It throws and error during the build because the CDK is looking for that directory.  This should remain.

schuettc commented on demo/meeting/front-end-resources/front-end.ts:
```diff
@@ -0,0 +1,83 @@
+ */
+export class FrontEnd extends Construct {
+    constructor(parent: Construct, name: string) {
+        super(parent, name);
+
+        // Content bucket
+        const siteBucket = new s3.Bucket(this, 'SiteBucket', {
+            websiteIndexDocument: 'index.html',
+            websiteErrorDocument: 'error.html',
```
No.  That should be changed to index.html as well.

schuettc commented on demo/meeting/front-end-resources/front-end.ts:
```diff
@@ -0,0 +1,83 @@
+            // the new bucket, and it will remain in your account until manually deleted. By setting the policy to
+            // DESTROY, cdk destroy will attempt to delete the bucket, but will error if the bucket is not empty.
+            removalPolicy: cdk.RemovalPolicy.DESTROY,
+            autoDeleteObjects: true, // NOT recommended for production code
```
That is correct.

schuettc commented:
Updated with many resolved issues from previous PR.  Cleaned up package.json and added DevPermissions.  Would like to wait on the changes to the join Lambda for further testing.

schuettc commented on demo/meeting/back-end-resources/back-end.ts:
```diff
@@ -0,0 +1,196 @@
+            role: lambdaChimeRole
+        });
+
+        const attendeeLambda = new lambda.Function(this, 'attendeeMeeting', {
```
Done

schuettc commented on demo/meeting/README.MD:
```diff
@@ -1,46 +1,23 @@
+- yalc [Installed](https://www.npmjs.com/package/yalc) - `npm i yalc -g'
+## Deploy
 
-## To run the demo app locally:
```
Instructions included to run locally.

schuettc commented on demo/meeting/back-end-resources/src/join.js:
```diff
@@ -0,0 +1,60 @@
+  }
+
+  console.info('Adding new attendee');
+  const attendeeInfo = (await chime.createAttendee({
+      MeetingId: meetingInfo.Meeting.MeetingId,
+      ExternalUserId: utils.uuid(),
+    }).promise());
```
Would like to wait on making a change here for future PR.

ltrung commented on demo/meeting/.gitignore:
```diff
@@ -1,2 +1,18 @@
+!/back-end-resources/lambda-layer/nodejs/node_modules/
+.DS_Store
+build/
+<<<<<<< HEAD
```
Seem like git conflict

ltrung commented on demo/meeting/back-end-resources/lambda-layer/nodejs/node_modules/utils.js:
```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: Apache-2.0
+
+var AWS = require('aws-sdk');
+var ddb = new AWS.DynamoDB();
```
Why is this under node_modules?

simmkyu replied:
From Court in the above comment:
> That's where it has to go. The directory structure for layers requires it.

ltrung commented on demo/meeting/back-end-resources/lambda-layer/nodejs/node_modules/utils.js:
```diff
@@ -0,0 +1,82 @@
+const attendeesTableName = process.env.ATTENDEES_TABLE_NAME;
+
+// Create a unique id
+exports.uuid = () => {
```
We want to get rid of this and use the uuid library

simmkyu commented on demo/meeting/back-end-resources/lambda-layer/nodejs/node_modules/utils.js:
```diff
@@ -0,0 +1,82 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
nit: Use `const` and `let` for consistency. Again, linters and formatters will help to fix minor syntax styles.

simmkyu commented on demo/meeting/back-end-resources/lambda-layer/nodejs/node_modules/utils.js:
```diff
@@ -0,0 +1,82 @@
+
+// Create a unique id
+exports.uuid = () => {
+  console.log("in uuid")
```
- Remove `console.log`s from this file?
- Or fix missing semicolon, single quote

simmkyu commented on demo/meeting/front-end-resources/front-end.ts:
```diff
@@ -0,0 +1,83 @@
+import s3deploy = require('@aws-cdk/aws-s3-deployment');
+import cdk = require('@aws-cdk/core');
+import { Construct } from '@aws-cdk/core';
+import iam = require('@aws-cdk/aws-iam')
```
- Missing semicolon. We can auto-fix these styles with formatters.

simmkyu reviewed (commented):
CDK is much more readable than the CF template. Thanks for making this fix!

I just noticed that I left some stying comments on files that you just moved from the existing repository. You can ignore them. We will need to integrate linters and formatters into the server side.

schuettc commented:
Closing to revise and resubmit.
